### PR TITLE
HIVE-24258: [CachedStore] Data miss match between cachedstore and rawstore

### PR DESCRIPTION
**Pull request by ashish-kumar-sharma**
Data miss match between cachedstore and rawstore

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

test case  added

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStoreUpdateUsingEvents.java [hunk 1]
```diff
@@ -419,18 +412,14 @@ public void testConstraintsForUpdateUsingEvents() throws Exception {
...
-    Assert.assertEquals(rawStoreConstraints.getCheckConstraints(), checkConstraints);
-    Assert.assertEquals(rawStoreConstraints.getForeignKeys(), foreignKeys);
+    SQLAllTableConstraints cachedStoreConstraints = new SQLAllTableConstraints();
+    cachedStoreConstraints.setPrimaryKeys(sharedCache.listCachedPrimaryKeys(catName, dbName, tblName));
+    cachedStoreConstraints.setForeignKeys(sharedCache.listCachedForeignKeys(catName, dbName, tblName, null, null));
+    cachedStoreConstraints.setNotNullConstraints(sharedCache.listCachedNotNullConstraints(catName, dbName, tblName));
+    cachedStoreConstraints.setDefaultConstraints(sharedCache.listCachedDefaultConstraint(catName, dbName, tblName));
+    cachedStoreConstraints.setCheckConstraints(sharedCache.listCachedCheckConstraint(catName, dbName, tblName));
+    cachedStoreConstraints.setUniqueConstraints(sharedCache.listCachedUniqueConstraint(catName, dbName, tblName));
+    Assert.assertEquals(rawStoreConstraints,cachedStoreConstraints);
```
Comment (sankarh) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStoreUpdateUsingEvents.java:422]: nit: Add space after ,

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 2]
```diff
@@ -2255,121 +2257,61 @@ private void create_table_core(final RawStore ms, final CreateTableRequest req)
...
+        if (CollectionUtils.isEmpty(constraints.getPrimaryKeys()) && CollectionUtils.isEmpty(constraints.getForeignKeys())
+                && CollectionUtils.isEmpty(constraints.getUniqueConstraints())&& CollectionUtils.isEmpty(constraints.getNotNullConstraints())&& CollectionUtils.isEmpty(constraints.getDefaultConstraints())
+            && CollectionUtils.isEmpty(constraints.getCheckConstraints())) {
           ms.createTable(tbl);
         } else {
           // Check that constraints have catalog name properly set first
-          if (primaryKeys != null && !primaryKeys.isEmpty() && !primaryKeys.get(0).isSetCatName()) {
-            for (SQLPrimaryKey pkcol : primaryKeys) pkcol.setCatName(tbl.getCatName());
+          if (CollectionUtils.isNotEmpty(constraints.getPrimaryKeys()) && !constraints.getPrimaryKeys().get(0).isSetCatName()) {
+            for (SQLPrimaryKey pkcol : constraints.getPrimaryKeys()) pkcol.setCatName(tbl.getCatName());
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:2267]: nit: Use for () { 
--
} 
even for single statement.
Or use this instead: constraints.getPrimaryKeys().forEach(pk -> pk.setCatName(tbl.getCatName()));

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 3]
```diff
@@ -2785,31 +2696,23 @@ public void add_not_null_constraint(AddNotNullConstraintRequest req)
     @Override
     public void add_default_constraint(AddDefaultConstraintRequest req)
         throws MetaException, InvalidObjectException {
-      List<SQLDefaultConstraint> defaultConstraintCols= req.getDefaultConstraintCols();
-      String constraintName = (defaultConstraintCols != null && defaultConstraintCols.size() > 0) ?
-          defaultConstraintCols.get(0).getDc_name() : "null";
+      List<SQLDefaultConstraint> defaultConstraints= req.getDefaultConstraintCols();
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:2699]: nit: Add space before =

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RawStore.java [hunk 4]
```diff
@@ -1499,20 +1499,11 @@ SQLAllTableConstraints getAllTableConstraints(String catName, String dbName, Str
   /**
    * Create a table with constraints
    * @param tbl table definition
-   * @param primaryKeys primary key definition, or null
-   * @param foreignKeys foreign key definition, or null
-   * @param uniqueConstraints unique constraints definition, or null
-   * @param notNullConstraints not null constraints definition, or null
-   * @param defaultConstraints default values definition, or null
    * @return list of constraint names
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RawStore.java:1502]: nit: Update comments for change in return type for all apis.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java [hunk 5]
```diff
@@ -58,14 +59,11 @@ public static String buildDbKeyWithDelimiterSuffix(String catName, String dbName
...
-    if (partVals == null || partVals.isEmpty()) {
-      return "";
-    }
-    return String.join(delimit, partVals);
+    return CollectionUtils.isNotEmpty(partVals) ? String.join(delimit, partVals) : "";
   }
 
   public static String buildTableKey(String catName, String dbName, String tableName) {
-    return buildKey(catName.toLowerCase(), dbName.toLowerCase(), tableName.toLowerCase());
+    return buildKey(StringUtils.normalizeIdentifier(catName),StringUtils.normalizeIdentifier(dbName),StringUtils.normalizeIdentifier(tableName));
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java:66]: nit: Space after ,

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 6]
```diff
@@ -2846,31 +2846,28 @@ public SQLAllTableConstraints getAllTableConstraints(String catName, String dbNa
...
   }
 
-  @Override public List<String> createTableWithConstraints(Table tbl, List<SQLPrimaryKey> primaryKeys,
-      List<SQLForeignKey> foreignKeys, List<SQLUniqueConstraint> uniqueConstraints,
-      List<SQLNotNullConstraint> notNullConstraints, List<SQLDefaultConstraint> defaultConstraints,
-      List<SQLCheckConstraint> checkConstraints) throws InvalidObjectException, MetaException {
-    List<String> constraintNames = rawStore
-        .createTableWithConstraints(tbl, primaryKeys, foreignKeys, uniqueConstraints, notNullConstraints,
-            defaultConstraints, checkConstraints);
+  @Override public SQLAllTableConstraints createTableWithConstraints(Table tbl, SQLAllTableConstraints constraints) throws InvalidObjectException, MetaException {
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2849]: nit: Shall keep annotation and api declaration ins separate lines.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java [hunk 7]
```diff
@@ -1568,12 +1568,7 @@ public void testPrimaryKeys() {
     List<SQLPrimaryKey> cachedKeys = sharedCache.listCachedPrimaryKeys(
             DEFAULT_CATALOG_NAME, tbl.getDbName(), tbl.getTableName());
 
-    Assert.assertEquals(cachedKeys.size(), 1);
-    Assert.assertEquals(cachedKeys.get(0).getPk_name(), "pk1");
-    Assert.assertEquals(cachedKeys.get(0).getTable_db(), "db");
-    Assert.assertEquals(cachedKeys.get(0).getTable_name(), tbl.getTableName());
-    Assert.assertEquals(cachedKeys.get(0).getColumn_name(), "col1");
-    Assert.assertEquals(cachedKeys.get(0).getCatName(), DEFAULT_CATALOG_NAME);
+    Assert.assertEquals(origKeys,cachedKeys);
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:1571]: nit: Space after ,

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStoreUpdateUsingEvents.java (hunk 1 above)
Comment (ashish-kumar-sharma) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStoreUpdateUsingEvents.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 2 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 3 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java (hunk 5 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java]: Done

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java (hunk 7 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RawStore.java (hunk 4 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RawStore.java:1503]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 6 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java]: done

**Review by sankarh — APPROVED** [review]
+1, LGTM

**Pull request merged**